compiler/rustc_hir_analysis/src/check/errs.rs

```diff
@@ -2,24 +2,58 @@ use rustc_hir as hir;
 use rustc_hir_pretty::qpath_to_string;
 use rustc_lint_defs::builtin::STATIC_MUT_REFS;
 use rustc_middle::ty::{Mutability, TyCtxt};
+use rustc_span::symbol::Ident;
 use rustc_span::Span;
 
 use crate::errors;
 
 /// Check for shared or mutable references of `static mut` inside expression
-pub fn maybe_expr_static_mut(tcx: TyCtxt<'_>, expr: hir::Expr<'_>) {
-    let span = expr.span;
-    let hir_id = expr.hir_id;
-    if let hir::ExprKind::AddrOf(borrow_kind, m, expr) = expr.kind
-        && matches!(borrow_kind, hir::BorrowKind::Ref)
-        && let Some(var) = path_if_static_mut(tcx, expr)
-    {
-        handle_static_mut_ref(tcx, span, var, span.edition().at_least_rust_2024(), m, hir_id);
+pub fn maybe_expr_static_mut(tcx: TyCtxt<'_>, expr: &hir::Expr<'_>) {
+    let mut span = expr.span;
+    let mut hir_id = expr.hir_id;
+    match expr.kind {
+        hir::ExprKind::AddrOf(borrow_kind, m, expr)
+            if matches!(borrow_kind, hir::BorrowKind::Ref)
+                && let Some(var) = path_if_static_mut(tcx, expr) =>
+        {
+            let sugg = if span.from_expansion() {
```

Comment:
Can you add a test case like this:
```rust
static mut FOO: (u32, u32) = (1, 2);

macro_rules! bar {
    ($x:expr) => {
        &mut ($x.0)
    };
}

unsafe {
    let _x = bar!(FOO);
}
```
and make sure it works as expected?

Reply:
test case has been added, and works as expected.

Reply:
🤔 it seems a bit arbitrary to me that we would strip exactly one nesting level here. I think we should make this clearer:
- `hir_id` is the id used to determine if the lint is enable there or not. Let's call it `lint_level_hir_id` everywhere, to make that clear. I don't think there is a reason to use anything other than the outer expression's HirId here.
- rename `sugg` to `suggest_addr_of` here too, so we know what it is about.
- `span` is the span highlighted in the error message, let's call it `err_span` to make that clearer. 
- we should probably determine `err_span` in `path_if_static_mut` to make this consistent everywhere.
- Depending on `err_span` it seems that sometimes the `format!()` case works and sometimes it just doesn't emit an error at all. But I have not figured out what the criteria are exactly.
 It seems a bit troublesome that emitting an error sometimes silently does nothing, based on what span is passed in.